src/hotspot/share/gc/shenandoah/shenandoahFullGC.cpp

```diff
@@ -458,14 +459,19 @@ class ShenandoahPrepareForGenerationalCompactionObjectClosure : public ObjectClo
                                                           ShenandoahHeapRegion* young_to_region, uint worker_id) :
       _preserved_marks(preserved_marks),
       _heap(ShenandoahHeap::heap()),
+      _tenuring_threshold(0),
       _empty_regions(empty_regions),
       _empty_regions_pos(0),
       _old_to_region(old_to_region),
       _young_to_region(young_to_region),
       _from_region(nullptr),
       _old_compact_point((old_to_region != nullptr)? old_to_region->bottom(): nullptr),
       _young_compact_point((young_to_region != nullptr)? young_to_region->bottom(): nullptr),
-      _worker_id(worker_id) {}
+      _worker_id(worker_id) {
+    if (_heap->mode()->is_generational()) {
+      _tenuring_threshold = _heap->age_census()->tenuring_threshold();
```

Comment:
Is this updated in the most recent stop-world marking, or is it using the one from the most recent concurrent collection? Could an interrupted concurrent marking cause a bogus census to be used here?

Partial censuses should be expunged.

Reply:
It is true that cancellations and partial censuses could lead to bogus partial censuses being used. I'll locate the right place in the cancellation path(s) where partial censuses will  need to be expunged. Essentially, these would be the points where concurrent marking is canceled (when censuses are conducted during the marking phase) and where concurrent evacuation is canceled (when censuses are conducted during the evacuation phase). All other cancellations are fine because the census data will be valid despite cancellation in those cases.

Reply:
In the event of degeneration, if a census-collection phase is interrupted mid-stream and restarted afresh, the partial census will need to be expunged. It'll then be replaced by the new census (in e.g. degenerate GC's mark or evac). In the case of a degenerate STW GC being cancelled and degenerating further to a full STW GC, during _its_ census-collection phase the partial census will be expunged. Any completed censuses from any census taking phase that persisted will be used as before. Currently, all calculations are independent of the (start and end) time of the (completed) censuses being used, but a future enhancement might potentially consider a temporal dimension. We'll capture the times in the census today, but those are not used by the algorithm today.

Reply:
The time tracking of the census is a potential future enhancement.

Today each new census clears any old partial or interrupted censuses, but we do not track the start and end times of completed censuses.